42/LOGOS-REPUTATION-INKINGUT (#524)
vacp2p/rfc: content/docs/rfcs/42/README.md

---
slug: 42
title: 42/LOGOS-REPUTATION-INKINGUT
name: Logos Node Reputation 
status: raw

> category: Informational

Reviewer: If this actually specifies the algorithm, it should be a `Standards Track` RFCs.

tags: logos/reputation
editor: Mark Evenson <mark.evenson@status.im>
created: 04-AUG-2022
revised: <2022-08-16 Tue 07:13>
contributors:
    - Álvaro Castro-Castilla
---

# Abstract

> We present Ikingut: an algorithm for purely local computation of peer

Reviewer: ```suggestion
This document specifies Ikingut, an algorithm for purely local computation of peer
```

Reviewer: In case we go with `Standards Track`.

node reputation based on their observed performance of a shared
computation.

# Background

The following document describes the Ikingut reputation algorithm,
designed to be used to augment the Logos consensus layer for various
purposes.

The use of objectives of defining a system for ongoing maintainance of
node reputation is to
 
   - Strengthen the Byzantine Fault Tolerance properties of the
     consensus mechanism, possibly making the algorithm more resilient
     to a higher number of byzantine nodes, or more aggressive
     strategies.

   - Providing a mechanism for bootstrapping networks, with a low
     number of nodes.

> # Theory 

Reviewer 2: How does this differ from Eigentrust at a high level in terms of properties?

## Ikingut Algorithm

> A deceptively simple algorithm for computing reputation of a other

Reviewer: ```suggestion
A deceptively simple algorithm for computing reputation of other
```

network nodes based on observed past performance.  

### Concept

The Ikingut algorithm for reputation is a local heuristic computed
from a single source of information: *how many correct responses did a
node give to queries*. The challenge in this metric is the definition
of *correct response*. We define a correct response as the decision

> the node achieves on the bit value (by means of Glacier).

Reviewer: I'd link the Glacier RFC here. 

... this sounds like it contradicts the independence of the reputation layer from Glacier.
Maybe:
As an example: when using Ikingut with Glacier, we define [...]

Reviewer: ~~(Maybe I misread the above and it is not meant to be used outside of Glacier?)~~
Line 65-66 say it is generic. So my comment above applies.

> This in essence summarizes the algorithm. There are implementation

Reviewer: implementation details should not be in the RFC anyways.
Maybe just: "details and parameters"

> This in essence summarizes the algorithm. There are implementation

Reviewer: > This in essence summarizes the algorithm.

Imo, a summary would require a more context on 

> We define a correct response as the decision the node achieves on the bit value 

How do you evaluate the decision a node achieves? 
(Also: does node refer to the responding node or the evaluating node?)

details and parameters that require some elaboration, as we will see
below in further detail. Later on we will provide an experimental
background to the specific design decisions.

Other relevant characteristics of the algorithm are:

- It’s simple. The complete implementation can be described in two
  functions.
- Lightweight. It doesn’t perform long computations nor use a lot of
  memory.
- Pluggable into Glacier [38/LOGOS-CONSENSUS-GLACIER][] and
  Snowball. Inkingut is designed to fit well with swarm-like consensus
  algorithms. Moreover, it augments the consensus algorithm without
  altering it.
- Adaptive. It processes network and behavior changes fast.
- Dynamic. It introduces new nodes consistently and fairly.
- It is robust against strategic adversaries. We will see these in more detail below.

> - No transitive trust. This avoids collusion, greatly simplifies. The

Reviewer: What does it simplify greatly?

>   tradeoff is more time to build the reputation list.

Reviewer: tradeoff with respect to what?

- Reasonable time to bootstrap reputation. With 10.000 nodes and k ~=
  9 is in the order of 8-10 min.

> ### Algorithm

Reviewer: In case we move this `Standards Track`, this should be a top-level section.
I'd name it `# Algorithm Specification` to make it explicit that this is the actual spec.

```

**Min-multiplicative reputation punishment**

The idea behind using a min function is that the punishment is much
harsher if the current reputation of the node is high. The min
function will select the highest reduction of reputation resulting
from these two operations:

- Multiplicative. This provides resiliency against patient attackers
  that have built reputation long-term. Some back-of-the-envelope
  calculations:
    - In one year at 1000 tx/s (network size: 10000 nodes, k=9), the
      expectation is around 0.9 req/s to each node. In 31536000
      secs/year, the max positive score: 28,382,400 (all
      reqs/node/year).
    - With a 2x punishment: 25 reqs to lose of all reputation
    - With a 3x punishment: 16 reqs to lose of all reputation
    
    Note that these requests don’t need to be consecutive, just close
    enough in time for the multiplicative punishment to take effect.
    
- Linear. This ensures the following properties:
    - Allow for new nodes to have a neutral score of zero.
    - Bring the opportunity of building reputation faster.
    - Ensure accumulation of negative scores for misbehaving nodes.
    

> # Implementation status

Reviewer: I'd remove this and not track implementation status in the  spec.

Incomplete.  No publically known implementations.  

> # Sovereignty Considerations

Reviewer: Imo, we should stick to the naming "Security Considerations", to be consistent with respect to other RFCs.
